Add Safe Abstractions for `FuriStreamBuffer`

cptpiepmatz opened the pull request:
Hey 👋🏻,
In this PR, I added some safe abstractions for the `FuriStreamBuffer`.

While working on it, I came across this note in the stream buffer implementation:
```
 * ***NOTE***: Stream buffer implementation assumes there is only one task or
 * interrupt that will write to the buffer (the writer), and only one task or
 * interrupt that will read from the buffer (the reader).
```

So, I figured, why not enforce this constraint in the type system? The result is a split into a sender and receiver, similar to the `mpsc` channel in the standard library. They implement `Send`, so they can be moved between threads, but not `Clone` or `Sync`, making sure we only have one of each in a pair. Unfortunately, I had to use `Arc` to give ownership to both sides, which means this requires `alloc`. Other than that, I’m happy with how the API turned out.

I also added an example where data is sent between different threads.

JarvisCraft commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,412 @@
+    }
+}
+
+unsafe impl Send for Sender {}
```
Safety comment would be useful here

cptpiepmatz replied:
I don't know what to really say here. Usually you would argue why `Send` is sound at this point but that's not really the reason that this type implements `Send`. Maybe I need to write quite some paragraph about the `Send` here, just like in `StreamBuffer`.

Not sure.

cptpiepmatz replied:
I made `StreamBuffer` `Send` and `Sync` with some comments why that is safe. For that I marked `send` and `receive` of the `StreamBuffer´  as `unsafe`.

JarvisCraft commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,412 @@
+    }
+}
+
+unsafe impl Send for Receiver {}
```
Same here

JarvisCraft reviewed (changes requested):
The API looks great. Here is the first iteration of nitpicking, but generally the PR looks super accurate

JarvisCraft commented:
It will take me some time to think on the `Arc`-related part, but it looks good for the start

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,412 @@
+/// stream buffer.
+/// Also both are [`Send`] to send them to other threads but not [`Sync`] to ensure that only one
+/// thread can use them at any time respectively.
+#[derive(Debug)]
```
Did that, the `NonNull` is handled by using the raw pointer for the field.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,412 @@
+    buffer_ref: Arc<StreamBuffer>,
+}
+
+impl Debug for Receiver {
```
Did that, the `NonNull` is handled by using the raw pointer for the field.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,412 @@
+    fn new(size: NonZeroUsize, trigger_level: usize) -> Self {
+        unsafe {
+            Self(NonNull::new_unchecked(sys::furi_stream_buffer_alloc(
+                size.into(),
```
I replaced them with local variables.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,412 @@
+        match updated {
+            true => Ok(()),
+            false => Err(()),
+        }
```
I dislike it visually with the if-else but I'm ok with it.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,412 @@
+        buffer_ref: stream_buffer.clone(),
+    };
+    let receiver = Receiver {
+        buffer_ref: stream_buffer.clone(),
```
Removed that one.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,412 @@
+///
+/// The `trigger_level` defines the number of bytes that must be present in the stream buffer
+/// before any blocked tasks waiting for data can proceed.
+// TODO: investigate what happens when we set the trigger level to high here
```
I just tried it out, it doesn't seem that the flipper checks if the `trigger_level` is stupid in the alloc. 🤷🏻

cptpiepmatz commented on crates/flipperzero/src/furi/mod.rs:
```diff
@@ -4,6 +4,8 @@ pub mod io;
 pub mod message_queue;
 pub mod rng;
+#[cfg(feature = "alloc")]
+pub mod stream_buffer;
```
It feels really bloated to a feature just because of the `Arc` I used. I could replace it with some custom counting logic using an `AtomicUsize` and `sys::malloc` to get a pointer but I'm not sure if that is better.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,412 @@
+
+    fn send(&self, data: &[u8], timeout: u32) -> usize {
+        unsafe {
+            sys::furi_stream_buffer_send(self.0.as_ptr(), data.as_ptr().cast(), data.len(), timeout)
```
I moved them all into local variables.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,412 @@
+    buffer_ref: Arc<StreamBuffer>,
+}
+
+impl Debug for Sender {
```
I did that.

dcoles commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,439 @@
+/// threads.
+/// However, they explicitly do not implement [`Sync`] or [`Clone`] to guarantee that only one
+/// writer and one reader can exist at any given time per task.
+pub fn stream_buffer(size: NonZeroUsize, trigger_level: usize) -> (Sender, Receiver) {
```
Perhaps `stream` might be a better name, since you're not actually getting a `StreamBuffer`—you're getting an abstraction on top of it.

cptpiepmatz replied:
This is replaced with `StreamBuffer::into_stream` now.

dcoles reviewed (changes requested):
Hi @cptpiepmatz!

Thanks for the PR! I really like the `Sender` / `Receiver` concept and how extensively you've documented everything 💖.

I think it would be useful to expose the low-level `StreamBuffer` API for `no-alloc` cases, but that's something we can certainly expose after merging this PR. Vast majority of comments are just little things for consistency with the other parts of the `flipperzero-rs` API.

dcoles commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,439 @@
+/// Basic Result which only indicates whether something succeeded or not.
+type EmptyResult = Result<(), ()>;
+
+// https://github.com/flipperdevices/flipperzero-firmware/blob/b723d463afccf628712475e11a3d4579f0331f5c/furi/core/base.h#L13
+const FURI_WAIT_FOREVER: u32 = 0xFFFFFFFF;
```
It would probably be good to have this in `furi::time`. There is a `MAX_DURATION_TICKS` (`0x7FFFFFFFF`) though that has a slightly different purpose.

cptpiepmatz replied:
I added `WAIT_FOREVER` as a new constant.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,439 @@
+        }
+    }
+
+    fn send(&self, data: &[u8], timeout: u32) -> usize {
```
I replaced that.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,439 @@
+    /// Further user reference is explained at the [`stream_buffer`] function.
+    fn new(size: NonZeroUsize, trigger_level: usize) -> Self {
+        let size: usize = size.into();
+        let ptr =
```
Added that safety comment.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,439 @@
+
+    fn send(&self, data: &[u8], timeout: u32) -> usize {
+        let self_ptr = self.0.as_ptr();
+        let data_ptr: *const c_void = data.as_ptr().cast();
```
Yep, removed the specific type.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,439 @@
+        unsafe { sys::furi_stream_buffer_is_empty(self_ptr) }
+    }
+
+    fn reset(&self) -> EmptyResult {
```
Is a `furi::Result` now, I also removed the `EmtpyResult` as it became obsolete.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,439 @@
+    fn drop(&mut self) {
+        // SAFETY:
+        // Since we keep an Arc in both, the sender and receiver, we know we only drop when both of
+        // them are dropped too.
```
That is true, I removed that one.

cptpiepmatz commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,439 @@
+    }
+}
+
+/// Check or modify the underlying furi stream buffer.
```
I handled this by adding a `as_stream_buffer` and `into_stream_buffer` methods on the `Sender` and `Receiver`.

cptpiepmatz commented:
I made the `StreamBuffer` public with all of its methods, the `send` and `receive` methods are `unsafe` as they need to uphold the guarantee that only writer and one reader can exists at a time.

The `Sender` and `Receiver` is moved in the private `stream` module. I re-exported the entire module but that allowed to just mark it as `#[cfg(feature = "alloc")]`. For that purpose I also removed the `stream-buffer´ feature again. I like this setup now, you can use the `StreamBuffer` without `alloc` but if you don't want to manage `send` and `receive` manually, you can use `Sender` and `Receiver` by enabling the `alloc` feature.

Since the `StreamBuffer` itself is now public, I added some methods on the `Sender` and `Receiver` to access the `StreamBuffer` and removed all the redirect methods. I also heavily updated the documentation to incorporate the new API.

dcoles commented:
Hi @cptpiepmatz. This looks great! I'd be happy to merge this once @JarvisCraft has taken a final look.

cptpiepmatz commented:
@dcoles could you please approve the workflow again? I added a `SetTriggerLevelError` to make clippy happy

cptpiepmatz commented:
@JarvisCraft is this ready to merge?

JarvisCraft commented:
> @JarvisCraft is this ready to merge?

Hi there, apologies for the long reply. I will take a look at the PR this week.

JarvisCraft commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,500 @@
+    #[cfg_attr(feature = "alloc", doc = "[`into_stream`](Self::into_stream)")]
+    /// which is available using the `alloc` feature.
+    pub fn new(size: NonZeroUsize, trigger_level: usize) -> Self {
+        let size: usize = size.into();
```
Non-blocker nit (ptobably, for the future changes): given that type annotation is already used, `from` seems more appropriate:

```suggestion
        let size = usize::from(size);
```

JarvisCraft commented on crates/flipperzero/src/furi/stream_buffer.rs:
```diff
@@ -0,0 +1,500 @@
+    use core::{cell::Cell, marker::PhantomData};
+
+    /// A zero-sized type used to mark types as `!Sync`.
+    type PhantomUnsync = PhantomData<Cell<()>>;
```
For future work on our side: this should probably be moved to some internal `utils` module

JarvisCraft reviewed (approved):
Apologies for the delay on code-review.

The PR looks amazing and the detailed docs are just the chef's kiss.

Thanks a lot for your contribution <3

JarvisCraft commented:
Squashed the commits so that `fix`-named  intermediate comments don't end up in `main`
